domain/application/state/application_endpoint.go

```diff
@@ -21,6 +21,118 @@ import (
+       ae.space_uuid AS &applicationEndpointBinding.space_uuid,
+       cr.name AS &applicationEndpointBinding.endpoint_name
+FROM   application_endpoint ae
+JOIN   charm_relation cr ON cr.uuid = ae.charm_relation_uuid
+JOIN   application a ON a.uuid = ae.application_uuid
+`, applicationEndpointBinding{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	extraEndpointBindingsStmt, err := st.Prepare(`
+SELECT a.name AS &applicationEndpointBinding.application_name,
+       aee.space_uuid AS &applicationEndpointBinding.space_uuid,
+       ceb.name AS &applicationEndpointBinding.endpoint_name
+FROM   application_extra_endpoint aee
+JOIN   charm_extra_binding ceb ON ceb.uuid = aee.charm_extra_binding_uuid
+JOIN   application a ON a.uuid = aee.application_uuid
+`, applicationEndpointBinding{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	defaultSpacesStmt, err := st.Prepare(`SELECT &applicationSpaceUUID.* FROM application`, applicationSpaceUUID{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	spacesStmt, err := st.Prepare(`SELECT &space.* FROM space`, space{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var allEndpointBindings []applicationEndpointBinding
+	var defaultSpaces []applicationSpaceUUID
+	var allSpaces []space
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var endpointBindings []applicationEndpointBinding
+		err := tx.Query(ctx, endpointBindingsStmt).GetAll(&endpointBindings)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("getting all endpoint bindings: %w", err)
+		}
+
+		var extraEndpointBindings []applicationEndpointBinding
+		err = tx.Query(ctx, extraEndpointBindingsStmt).GetAll(&extraEndpointBindings)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("getting all endpoint bindings: %w", err)
+		}
+
+		err = tx.Query(ctx, defaultSpacesStmt).GetAll(&defaultSpaces)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("getting all default spaces: %w", err)
+		}
+
+		err = tx.Query(ctx, spacesStmt).GetAll(&allSpaces)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("getting all spaces: %w", err)
+		}
+
+		allEndpointBindings = append(endpointBindings, extraEndpointBindings...)
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	spaceUUIDToName := make(map[network.SpaceUUID]network.SpaceName)
+	for _, space := range allSpaces {
+		spaceUUIDToName[space.UUID] = space.Name
+	}
+
+	result := make(map[string]map[string]network.SpaceName)
+	for _, space := range defaultSpaces {
+		if _, ok := result[space.ApplicationName]; !ok {
+			result[space.ApplicationName] = make(map[string]network.SpaceName)
+		}
+		spaceName, ok := spaceUUIDToName[space.SpaceUUID]
+		if !ok {
+			return nil, errors.Errorf("space %q not found", space.SpaceUUID)
+		}
+		result[space.ApplicationName][""] = spaceName
```

Comment:
I wish we used a constant for the empty wildcards.

Reply:
Me too!